Search package: Add search auto_config CLI

kangzj opened the pull request:
#### Changes proposed in this Pull Request:
The PR proposes a `auto_config` CLI to allow the functionality to be used thru WPCLI. The functionality for now is only intended for the search E2E tests.

#### Jetpack product discussion
n/a

#### Does this pull request change what data or activity we track or use?
no

#### Testing instructions:
* Go to a site with Jetpack Search subscription
* Use a non block theme (Block themes are supported thru #22613 )
* Enable Instant Search
* Remove search widgets from sidebars
* Remove overlay sidebar wigets
* Run `jetpack wp jetpack-search auto_config`
* Ensure the removed widgets are added again.



![image](https://user-images.githubusercontent.com/1425433/152264248-a3da6bd2-bac1-4ece-a101-81c83d7fed3b.png)

kangzj committed:
added jetpack search auto config cli

kangzj committed:
changelog

kangzj committed:
use Exception instead of Throwable

kangzj committed:
extracted auto_search_no_pri for cli

a8ck3n commented:
I had a go at testing this but couldn't get it to work. The steps I took were:

1. Changed my site theme to Twenty Sixteen.
2. Removed the search widgets from the sidebar.
3. Attempted to run the new CLI command (via Docker env, per [this](https://github.com/Automattic/jetpack/blob/master/tools/docker/README.md#using-wp-cli)).
4. Received the following error:

> Error: Jetpack Search: auto config has run already. You could add option `--force` to force run.

So at a minimum, the new code path is running but it's not seeing the `--force` argument. I tried running it the following ways:

- `jetpack docker wp jetpack-search auto_config --force`
- `jetpack docker wp jetpack-search auto_config -- --force`
- `jetpack docker wp jetpack-search auto_config - --force`
- `jetpack docker wp jetpack-search auto_config -- -force`

So I'm assuming that flag is not being passed through `jetpack docker` correctly such that wp-cli sees it.

kangzj committed:
no need to return, as error exit the command

kangzj committed:
Merge branch 'master' into add/search-wp-cli-auto-config

kangzj committed:
remove force option, because auto config function has some conditions inside which would prvevent it from running multiple times

kangzj commented:
Hi @a8ck3n Thanks for testing and instructions to replicate the issue. I had a go today, and can confirm you were right. The force option doesn't work, because there are some mechanism inside the auto config function which stops it from running multiple times. And the correct way to pass `--force` is:

`jetpack docker wp jetpack-search auto_config -- --force`

But which doesn't work for now. I removed the option and will revisit after we refactor the auto config function. Updated test instructions.

kangzj committed:
bump version

kangzj committed:
cast string to int for PHP 8

jsnmoon commented on projects/packages/search/src/instant-search/class-instant-search.php:
```diff
@@ -313,6 +313,15 @@ public function auto_config_search() {
+	 *
+	 * @since  8.3.0
+	 */
+	public function auto_config_search_no_pri() {
```
What does `no_pri` mean here?

Also, what do you think about overloading the original function call with a default parameter instead of creating a new function? Something like...

```php
	public function auto_config_search( $check_permissions = true ) {
		if ( $check_permissions && ! current_user_can( 'edit_theme_options' ) ) {
```

With this, we could invoke `Instant_Search::instance()->auto_config_search( false );` in the CLI.

kangzj replied:
Good point. There was an ancient action called `wp_ajax_no_priv_your_action`, but still it's is not descriptive.
Had another consideration, I would rather just remove the privileges check here, b/c the function are called from APIs which have privilege checks of their own, and ARE not the same as the check in here.

REF1: https://github.com/Automattic/jetpack/blob/301c0f53f31adf0ea54cebd58bcbe67b387fe662/projects/plugins/jetpack/_inc/lib/core-api/class.jetpack-core-api-module-endpoints.php#L1249
REF2: https://github.com/Automattic/jetpack/blob/301c0f53f31adf0ea54cebd58bcbe67b387fe662/projects/packages/search/src/class-rest-controller.php#L116   <- we probably want to widen the privilge check here

jsnmoon replied:
Sounds good!

jsnmoon reviewed (commented):
Works as expected! Left a comment inline.

kangzj committed:
remove unnecessary privileges check

kangzj committed:
remove permission check

kangzj committed:
Merge branch 'master' into add/search-wp-cli-auto-config
